### DISCUSSION TITLE
How to use fixture in a pytest TestClass with custom plugin

### DISCUSSION THREAD

**Discussion by sar009** [pytest-dev/pytest#11476] [Q&A]
I am unable to use a pytest fixture in a class-based test. My code looks like the following.

main.py
```python
#!/usr/bin/env python3
import inspect

import pytest


class MyPlugin:

    __name__ = 'my_plugin'

    def pytest_pycollect_makeitem(self, collector, name, obj):
        items = []
        if ((inspect.isfunction(obj) or inspect.ismethod(obj))
                and (name.startswith("test_") or name.endswith("_test"))):

            items.append(pytest.Function.from_parent(collector, name=name))

        if inspect.isclass(obj):
            items.append(pytest.Class.from_parent(collector, name=name))

        return items


if __name__ == '__main__':
    pytest.main(["test.py", "-s"], plugins=[MyPlugin()])
    print("------------------------------")
    pytest.main(["test.py", "-s"])
```

test.py

```python
import pytest


@pytest.fixture
def my_fixture():
    print(f"hello")


class TestClass:
    @pytest.mark.usefixtures("my_fixture")
    def test_in_class_with_decorator(self):
        assert True

    def test_in_class(self, my_fixture):
        assert True


def test_out_class(my_fixture):
    assert True
```

when running the code as ./main.py I am getting the error as

```shell
FAILED test.py::TestClass::test_in_class - TypeError: TestClass.test_in_class() missing 1 required positional argument: 'my_fixture'
```
The error happens when I am using the plugin(MyPlugin). What needs to be changed in the plugin for a class-based test to access the fixture as a test method parameter?

**Comment by RonnyPfannschmidt** [comment]
Fixture info isn't being added,

The own version in pytester Passes over a lot More

**Reply by sar009** [reply]
@RonnyPfannschmidt is there any example or internal class i can refer to

**Reply by RonnyPfannschmidt** [reply]
I'm not aware of external ones, the built-in python plugin has correct invocations

**Reply by sar009** [reply]
can you please point me to the plugin? essentially I am looking for the default behaviour of `pytest_pycollect_makeitem`

**Reply by sar009** [reply]
@RonnyPfannschmidt can you please point me to the plugin? essentially I am looking for the default behaviour of `pytest_pycollect_makeitem`

**Reply by The-Compiler** [reply]
A search for "python" in the repository [reveals a `python.py`](https://github.com/pytest-dev/pytest/blob/main/src/_pytest/python.py). Also it seems quite rude to me to post the exact same thing again an hour (!) after first posting it. People have lives outside of their screen too :wink:

**Reply by sar009** [reply]
I am a new user of Pytest and as a user of any library, I would interact with the provided APIs and not with internals if someone is reaching out for help with something that deals with the internals that means the documentation of the same was not helpful or the web search for the same didn't yield helpful result or tools like chatGPT was not able to help either or they were not able to figure out debugging the code. Again as a user, it was not apparent to me that there exists a plugin in the library named `python`. Maybe if in the first reply, the link of the file was mentioned we might not have gone through this.

While I am extremely sorry for commenting twice within an hour and also thankful for the prompt reply, I also believe you should have shown some empathy to the user because this might be the last door they are knocking on after trying everything possible, the user who is NOT a maintainer of the library.

**Reply by RonnyPfannschmidt** [reply]
Please reflect a little, he showed quite some empathy.

The solution was provided first and the behavior was called out as an impression, not as a fact.

**Reply by The-Compiler** [reply]
If you're writing a custom pytest plugin, you're interacting with pytest internals. That's somewhat unavoidable, and if you're a pytest beginner, that's not exactly what I'd start with. On the other side of the coin, people answering you will typically not assume that you're a beginner if you're using what's arguably the most advanced part of a framework.

I agree it would have been nice to have that link in @RonnyPfannschmidt's first answer, but look: You're talking to people who spend their free time helping you, and typically dozens of other people every single day. Above, you came across as extremely pushy when not getting a real-time answer from a maintainer, again, doing this in their free time. Empathy goes both ways.

**Reply by sar009** [reply]
Folks, I am not implying you owe me a reply, and that too in real-time. Yes, I tagged @RonnyPfannschmidt hoping to get his attention in case he is available YES that is pushy behavior(may be not extreme). 
I am well aware of the contribution you folks are making to society if it was not for great souls like you there wouldn't exist a framework like PyTest. My only request to you would be to draft replies that are helpful for the reader if the reply is a single liner like the ones above that would lead to more to and fro in the comments eventually wasting your time and not being helpful to the reader. This is the part where I say you should show empathy to the user who is not a maintainer and provide more context to them.
For instance, my query is not yet resolved even after referring to [pytest_pycollect_makeitem](https://github.com/pytest-dev/pytest/blob/main/src/_pytest/python.py#L232)

**Reply by sar009** [reply]
the following change has helped me but I am not sure if that is the right approach especially because I am accessing protected member to get fixtureinfo.
```python
    def pytest_pycollect_makeitem(self, collector, name, obj):
        items = []
        if ((inspect.isfunction(obj) or inspect.ismethod(obj))
                and (name.startswith("test_") or name.endswith("_test"))):

            fixture_info = pytest.Function.from_parent(collector, name=name, callobj=obj)._fixtureinfo
            items.append(pytest.Function.from_parent(collector, name=name, fixtureinfo=fixture_info))

        if inspect.isclass(obj) and collector.istestclass(obj, name):
            items.append(pytest.Class.from_parent(collector, name=name, obj=obj))

        return items
```
